### ISSUE TITLE
Expand optimized LoRA kernels to lm_head & embed_tokens

### ISSUE THREAD

**Issue by hcsolakoglu** [axolotl-ai-cloud/axolotl#2720]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Currently LoRA optimizations target only specific modules, but no optimized kernels exist for lm_head or embed_tokens. As a result full-model fine-tuning with lora suffers from reduced throughput and increased memory usage

Axolotl’s current LoRA setup already provides optimized kernels for :

```yaml
lora_mlp_kernel: true
lora_o_kernel:    true
lora_qkv_kernel:  true
lora_target_modules:
  - q_proj
  - k_proj
  - v_proj
  - o_proj
  - gate_proj
  - down_proj
  - up_proj
```

However, there are no analogous optimized kernels for:

* `lm_head`
* `embed_tokens`

When you fine-tune entire model, missing these kernels leads to significant drop in throughput and increased GPU memory usage.

### ✔️ Solution

Extend existing LoRA optimizations (eg via Triton or custom CUDA kernels) to cover lm_head and embed_tokens. Benchmarks should demonstrate restored training throughput and lower peak memory footprint. Once validated, integrate new kernels into Axolotl’s fine-tuning pipeline.

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by NanoCode012** [comment]
I wonder if this is something that we can extend to, or is it unrelated? @djsaunde 

In meantime, have you tried any of our optimizations for cross entropy? CCE  or Liger ? https://docs.axolotl.ai/docs/custom_integrations.html

**Comment by winglian** [comment]
there is little need for new kernels for `lm_head` as both liger FLCE or CCE solve for this.
@NanoCode012 for `embed_tokens`, this might get solved with the new PEFT feature we talked about @ https://huggingface.co/docs/peft/main/en/developer_guides/lora#efficiently-train-tokens-alongside-lora

EDIT: technically embed_tokens are lookup tables, and doesn't involve any large matmuls or activations. https://discuss.pytorch.org/t/how-does-nn-embedding-work/88518/14

**Comment by NanoCode012** [comment]
@winglian , agreed that `embed_tokens` are not the expensive operations.

For that linked feature, we can indeed add it, however, I'm not sure what's the most intuitive way for a user to use this. Do they manually pass in the indices of the token to train? Or, do they pass in the token_string? If the latter, we would then have to check the config and parse it into tokens.

**Comment by winglian** [comment]
I think we can leverage some our existing added tokens feature, between

basically any token ids from `added_tokens_overrides` and `fix_untrained_tokens`
